### DISCUSSION TITLE
add subtitle to a resource page?

### DISCUSSION THREAD

**Discussion by wivaku** [filamentphp/filament#6061] [Ideas]
It's my understanding it is currently not possible to add a subtitle to a resource page.

Sometimes it helps to add some clarification.

Example:

<img width="1281" alt="image" src="https://user-images.githubusercontent.com/2084872/228649068-5910c5d2-1648-4f5c-9fbb-8ea7674cbdfd.png">

E.g. 
```php
// CustomerResource

class CustomerResource extends Resource
{

  //protected static ?string $title = 'Custom Page Title';
  protected static ?string $subtitle = 'example of subtitle';
```

**Comment by danharrin** [comment]
`$subheading`

**Reply by arthurnmachado** [reply]
Is there any way to modify the subheading, such as its size or color?

**Reply by danharrin** [reply]
Using CSS in your theme file, yes

**Comment by wivaku** [comment]
Nice! 

Could not find this in the documentation, right?
Looked here:
https://filamentphp.com/docs/2.x/admin/pages/getting-started#customization

Having trouble to get it to work though (strange: same for title btw).

<img width="803" alt="image" src="https://user-images.githubusercontent.com/2084872/228780057-33ea0ba3-2cb9-44a6-a8ef-c7d17f07cf66.png">

![image](https://user-images.githubusercontent.com/2084872/228780235-60ced163-930a-4b82-afd8-0d7ed7bd044e.png)

**Reply by danharrin** [reply]
You need to put the properties on a page class, not a resource class

**Comment by wivaku** [comment]
That makes sense. To be sure, my question was about a resource.  
`$navigationLabel`, `$slug` do work there. But `$title` and `$subheading` don't.
Is it possible (or: can it be added in the future) to customize `$title` and add `$subheading` to a resource?

**Reply by danharrin** [reply]
the resource is to define stuff that is used across multiple pages. any page-specific things like heading are overridden per-page. resource labels are used across all pages so they belong in the resource

**Comment by wivaku** [comment]
Ah, lightbulb moment 💡 !
I was focused on `Resources\Shop\ProductResource.php`  
and I should have looked at `Resources\Shop\ProductResource\ListProducts.php`.

I have added to `Resources\Shop\ProductResource\ListProducts.php`:
```
    protected ?string $subheading = 'Custom subheading';
```

And it works. 👍 

Note that if I add `static` (like the `navigationLabel` and `slug` for a resource) I get an error message:  
`Cannot redeclare non static Filament\Pages\Page::$subheading as static`

Thanks Dan!

**Comment by bilogic** [comment]
Can subheadings be translated?

**Reply by danharrin** [reply]
override `getSubheading()`

**Reply by bilogic** [reply]
would it make sense to change filament to return `__($this->subheading);` always?

**Reply by danharrin** [reply]
No, we did automatic translation by default in v1 and encountered problems.

Also, using English translation strings 1:1 as keys for other languages is bad practice for your own maintainability and for contextual translations, so you shouldn't be doing it anyway.

**Reply by bilogic** [reply]
ok thanks

**Reply by bilogic** [reply]
Actually, what I typically do is to replace `$this->subheading` with a proper key if I'm doing translation. 

Meaning to say, while `$this->subheading` can start with pure English in `__($this->subheading)`, once I start translation, I will have an `en.json` and replace `$this->subheading` with some sort of key.

Would you be able to briefly share the problems of translations in v1 for my learning purpose?

**Reply by danharrin** [reply]
People didn't like the fact that things were being translated by default, not much else to add. They didn't like the fact that they had to override the method to be able to disable the feature when it feels like something that should be opt-in. And I agree with them

**Reply by bilogic** [reply]
Ok, but how did `__($this->subheading)` affect them if they don't need translations?

**Reply by danharrin** [reply]
They had translation files for common words, and didnt want them translated automatically.

**Reply by bilogic** [reply]
LOL, ok, in that case, may I propose something that will fit all our needs. It will still be opt in, but `getSubheading` does not need to be overridden in every class.
```PHP
public function getSubheading(): string | Htmlable | null
{
    if (config("filament.always_translate"))
    {
        return __($this->subheading);
    }
    
    return $this->subheading;
}
```


This proposal would not be needed if I could write `protected ?string $subheading = __("translate this");`, unfortunately it is illegal syntax.

**Reply by danharrin** [reply]
I think that would add too much maintenance burden ensuring that the setting is added in every place, I don't think I would accept that.

**Reply by bilogic** [reply]
What is your scope for "every place"? 

For a start, I think `getTitle()` and `getSubHeading()` will do, the rest can be added progressively.
I'm willing to do the PRs for any reported translation issues for this mechanism. 

To keep it maintainable, perhaps we can have this in trait to make it easy to remove (if there is ever a need). "Perhaps" because I have not looked at the consistency and location of the method names yet, but I think a trait would be very clean.

What kind of maintenance concerns do you have?

**Reply by danharrin** [reply]
Look, I'm not interested in the feature. Eventually every string will need to be added, it's just too much extra code. I'm happy with the current system of defining methods to pass translation strings.

**Reply by bilogic** [reply]
ok thanks

**Comment by BrilianPP** [comment]
Hello, please help me, 
in my program, i have error "Missing required parameter for [Route: filament.admin.resources.periodes.rakers.index] [URI: admin/periodes/{parent}/rakers] [Missing parameter: parent]."
please help me for solve this problem
big thanks
